File: Top-Interview-Questions/Array/Easy-Collection/1-two_sum.cpp
// First Solution: Brute Force
// Time Complexity: O(n^2)
// Space Complexity: O(1)
// Runtime: 1024 ms, faster than 5.04% 
// Memory Usage: 9.2 MB, less than 92.37%
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        
        vector<int> temp;

        for(int i = 0; i < nums.size(); i++){ // pick first number for addition
            
            for(int j = i+1; j < nums.size(); j++){ // pick second number for addition
                
                if(target == nums[j] + nums[i]){ 
                    temp.push_back(i);
                    temp.push_back(j);
                }
            }
        }
        
        return temp;
    }
};


// Second Solution: Map + Vector
// Time Complexity: O(n)
// Space Complexity: O(n)
// Runtime: 32 ms, faster than 39.09%
// Memory Usage: 13 MB, less than 5.17%
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        
        map<int, vector<int>> temp;
        vector<int> pair;
        
        for(int i = 0; i < nums.size(); i++){
            
            temp[nums[i]].push_back(i);
        }
        
        for(int i = 0; i < nums.size(); i++){
            
            if(temp[target-nums[i]].size()){
                
                if(target == 2*nums[i]){
                    if(temp[nums[i]].size() > 1){
                        pair.push_back(temp[nums[i]][0]);
                        pair.push_back(temp[nums[i]][1]);
                        return pair;
                    }
                } else if(temp[target-nums[i]].size()){
                    pair.push_back(temp[nums[i]][0]);
                    pair.push_back(temp[target-nums[i]][0]);
                    return pair;
                }
            }
        }
        
        return pair;
    }
};

File: Top-Interview-Questions/Array-Easy-Collection/1-two-sum.js
/*
    Initial naive solution
    Runtime: 60 ms, faster than 65.54% of JavaScript online submissions for Two Sum.
    Memory Usage: 37.1 MB, less than 8.68% of JavaScript online submissions for Two Sum.
 */
var twoSum = function(nums, target) {
    if (nums.length <= 1) {
        return null;
    }

    const map = new Map();

    nums.forEach((number, index) => {
        if (map.has(number)) {
            map.get(number).push(index);
        } else {
            map.set(number, [index]);
        }
    });

    return nums.reduce((result, current) => {
        if (result.length === 0) {
            const remainder = target - current;

            if (map.has(remainder)) {
                if (remainder !== current) {
                    result.push(map.get(current)[0]);
                    result.push(map.get(remainder)[0]);
                }
                else if (map.get(remainder).length > 1) {
                    result.push(map.get(current)[0]);
                    result.push(map.get(current)[1]);
                }
            }
        }

        return result;
    }, []);
};

/*
    Optimal solution
    Runtime: 56 ms, faster than 80.86% of JavaScript online submissions for Two Sum.
    Memory Usage: 36.7 MB, less than 10.33% of JavaScript online submissions for Two Sum.
 */
var twoSum = function(nums, target) {
    if (nums.length <= 1) {
        return null;
    }

    const map = new Map();

    for (const [index, value] of nums.entries()) {
        if (!map.has(value)) {
            map.set(target - value, index);
        } else {
            return [map.get(value), index];
        }
    };
}